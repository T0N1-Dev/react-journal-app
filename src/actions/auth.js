import { types } from "../types/types";
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, updateProfile } from "firebase/auth";
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const {user} = await signInWithEmailAndPassword(auth, email, password);

            dispatch( login(user.uid, user.displayName));
            dispatch(finishLoading());
        }
        catch (err) {
            dispatch(finishLoading());
            Swal.fire('Error', err.message, 'error');
        } 
    }
}

export const startRegisterWithEmailPassswordName = (email, password, name) => {

    return async (dispatch) => {
        try {
            dispatch(startLoading());
            const { user } = await createUserWithEmailAndPassword(auth, email, password);
            await updateProfile(user, { displayName: name });
            
            dispatch(
                login(user.uid, user.displayName)
            );
            dispatch(finishLoading());
        }
        catch (err) {
            Swal.fire('Error', err.message, 'error');
            dispatch(finishLoading())
        } 
    }
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
          const { user } = await signInWithPopup(auth, googleAuthProvider);
          dispatch(
            login( user.uid, user.displayName )
          );
        } catch (error) {
          console.error("Error al iniciar sesiÃ³n con Google:", error);
        }
    };
}

export const login = (uid, displayName) => ({
        type: types.login,
        payload: {
            uid,
            displayName
        }
    })

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await auth.signOut();
            dispatch( logout() );
            dispatch( noteLogout() );
        } catch (err) {
            console.error("Error al hacer Logout:", err);
        }
    }
}

export const logout = () => ({
    type: types.logout
})