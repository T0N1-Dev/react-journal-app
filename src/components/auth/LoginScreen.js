import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import validator from 'validator';
import { setError } from '../../actions/ui';

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const { mgError, loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'test@gmail.com',
    password: '123456',
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();

    if (validator.isEmail(email)) {
      dispatch(startLoginEmailPassword(email, password));
    } else {
      dispatch(setError('Invalid email: ' + email));
    }
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form 
        onSubmit={handleLogin}
        className='animate__animated animate__fadeIn animate__faster'
      >
        {mgError && <div className="auth__alert-error">{mgError}</div>}

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        <div className="auth__social-network">
          <p>Login with social networks</p>

          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQRB4kg1jgnqD5fEaCSHKMoUdFeTBj9CH_opA&s"
                alt="google button"
              />
            </div>

            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register" className="link" data-testid="create-link">
          Create new account
        </Link>
      </form>
    </>
  );
};

