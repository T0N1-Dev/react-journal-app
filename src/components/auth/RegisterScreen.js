import React from 'react'
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPassswordName } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { mgError, loading } = useSelector( state => state.ui );

  const [ { name, email, password, password2 }, handleInputChange, reset ] = useForm({
    name: 'Prueba', 
    email:'prueba@gmail.com', 
    password:'123456', 
    password2:'123456'
  });

  const handleRegister = (e) => {
    e.preventDefault();

    if ( isFormValid() ){
      dispatch(startRegisterWithEmailPassswordName(email, password, name));
    }
    reset();
  };

  const isFormValid = () => {

    if ( name.trim().length === 0 || !validator.isAlpha(name) ) {
      dispatch(setError('Invalid name: ' + name));
      return false;
    } else if ( email.trim().length === 0 || !validator.isEmail(email) ) {
      dispatch(setError('Invalid email: ' + email));
      return false;
    } else if ( password !== password2 || password.length < 6) {
      dispatch(setError('Invalid passwords'));
      return false;
    }

    dispatch( removeError() );
    return true;
  }

  return (
    <>
      <h3 className='auth__title'>Register</h3>

      <form 
        onSubmit={ handleRegister }
        className='animate__animated animate__fadeIn animate__faster'
      >

      {
        mgError && 
        (
          <div className='auth__alert-error'>
            { mgError }
          </div>
        )
      }

      <input 
          type='text'
          placeholder='Name'
          name='name'
          className='auth__input'
          autoComplete='off'
          value={name}
          onChange={handleInputChange}
        />  

        <input 
          type='text'
          placeholder='Email'
          name='email'
          className='auth__input'
          autoComplete='off'
          value={email}
          onChange={handleInputChange}
        />  

        <input 
          type='password'
          placeholder='Password'
          name='password'
          className='auth__input'
          value={password}
          onChange={handleInputChange}
        />

        <input 
          type='password'
          placeholder='Confirm Password'
          name='password2'
          className='auth__input'
          value={password2}
          onChange={handleInputChange}
        />

        <button
          type='submit'
          className='btn btn-primary btn-block mb-5'
          disabled={ loading }
        >
          Register
        </button>

        <Link
        to="/auth/login"
        className="link"
        >
          Already registered?
        </Link>
      </form>  
    </>
  )
}
