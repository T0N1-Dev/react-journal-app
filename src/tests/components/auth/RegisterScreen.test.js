import { configureStore } from "@reduxjs/toolkit";
import { authReducer } from "../../../reducers/authReducer";
import { uiReducer } from "../../../reducers/uiReducer";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { fireEvent, render } from "@testing-library/react";
import { types } from "../../../types/types";
import '@testing-library/jest-dom';

describe('Tests in <RegisterScreen/>', () => {
    let store;
    let testContainer;
    
    beforeEach(() => {
        store = configureStore({
            reducer: {
                auth: authReducer,
                ui: uiReducer
            },
            preloadedState: {
                auth: {},
                ui: {}
            }
        });

        const { container } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        store.dispatch = jest.fn();

        testContainer = container;
    });

    test('should match the snapshot', () => { 
        expect(testContainer).toMatchSnapshot();
    });

    test('should trigger setError with invalid email', () => {
        const input = testContainer.querySelector('input[name="email"]');
        fireEvent.change(input, { target: { value: 'test' } });

        const form = testContainer.querySelector('form');
        fireEvent.submit(form);

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith({
            type: types.uiSetError,
            payload: 'Invalid email: test'
        });
    });

    test('should show the error message', () => {
        store = configureStore({
            reducer: {
                auth: authReducer,
                ui: uiReducer
            },
            preloadedState: {
                auth: {},
                ui: {
                    loading: false,
                    mgError: 'Invalid email:'
                }
            }
        });

        const { container } = render(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        const errorMessage = container.querySelector('.auth__alert-error');
        expect(errorMessage).toBeInTheDocument();
        expect(errorMessage.textContent).toBe('Invalid email:');
    });
});