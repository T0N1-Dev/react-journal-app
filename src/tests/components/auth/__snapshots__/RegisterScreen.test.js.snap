// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <RegisterScreen/> should match the snapshot 1`] = `
<div>
  <h3
    class="auth__title"
  >
    Register
  </h3>
  <form
    class="animate__animated animate__fadeIn animate__faster"
  >
    <input
      autocomplete="off"
      class="auth__input"
      name="name"
      placeholder="Name"
      type="text"
      value="Test"
    />
    <input
      autocomplete="off"
      class="auth__input"
      name="email"
      placeholder="Email"
      type="text"
      value="test@gmail.com"
    />
    <input
      class="auth__input"
      name="password"
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      class="auth__input"
      name="password2"
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      class="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      class="link"
      href="/auth/login"
    >
      Already registered?
    </a>
  </form>
</div>
`;
